### ISSUE TITLE
render_on_canvas: true doesn't work when images aren't preloaded?

### ISSUE THREAD

**Issue by becky-gilbert** [jspsych/jsPsych#1305]
See #1304

**Comment by becky-gilbert** [comment]
When the image isn't preloaded, it may not be finished loading when the `ctx.drawImage` function is called, which means that nothing is drawn on the canvas. And unlike presenting images with an `img` tag, if there's a loading delay then the image will not automatically appear _after_ it has finished loading. We can fix this by making sure that the image isn't drawn on the canvas until after it has finished loading.

One option is to add an `onload` event handler to the image that calls `ctx.drawImage`. This would mean that the trial starts immediately and the image appears if/when it loads (as opposed to delaying the start of the trial until the image loads). This is what happens when the image is presented with an `img` tag, which means the plugin's behavior with non-preloaded images would be consistent regardless of whether `render_on_canvas` is `true` or `false`. @jodeleeuw does this approach sound ok to you?

**Comment by jodeleeuw** [comment]
Yeah, I like that solution!

Creating the preloading plugin (I suggest it in this thread #1234) will also help solve this issue by making it easier to preload small sets of images. I should open an issue for that plugin and add it to 6.3....

**Comment by becky-gilbert** [comment]
Great, thanks!

**Comment by becky-gilbert** [comment]
This issue is now fixed in the bugfix-image-canvas branch. I updated the image-* plugins so that they include an `img.onload` function that will draw the image on the canvas after it loads, assuming it isn't immediately available (i.e. cached). So now when the image isn't preloaded, it will be displayed as soon as it loads. Note that in practice this may mean that the image is never displayed, if it doesn't load before the trial ends.

Unfortunately it wasn't possible to rely solely on the `img.onload` function to draw the image on the canvas, because when the image is immediately available (preloaded), relying on the load event to draw it to the canvas produces a delay between when the trial starts and when the image is drawn (and the whole reason for adding this canvas rendering option was to get rid of this delay!). So the solution was to include the `img.onload` function _in addition to_ the previous method for drawing the image, rather than _in place of_ the previous method. If anyone has suggestions for a more elegant fix, I'd be happy to hear them!

I couldn't think of a way to add a test for this. This is an issue that ideally would've been caught in our existing tests that check that the image has been added to the screen, since these test trials do not use preloading (`auto_preload: false`). However I'm guessing that the image doesn't actually 'load' in Jest the way it does in a browser, which means that we're not catching problems that relate to image loading delays. And even if there's a way to simulate the image loading delay, currently all of our image-* plugin tests only use `render_on_canvas: false` because I haven't yet been able to figure out [canvas support in Jest](https://stackoverflow.com/questions/33269093/how-to-add-canvas-support-to-my-tests-in-jest). We could test whether the canvas element exists on the page, but that's not particularly helpful since what we really need to know whether or not an image has been drawn on the canvas. Again, I'd love to hear suggestions for this.
